### PR TITLE
feature: Add Open AI as model provider

### PR DESCRIPTION
**Pull request by EdwardTang**
#  Add Direct OpenAI API Support

## Overview
This PR adds direct OpenAI API support to the Jupyter AI Agents, allowing users to choose between Azure OpenAI and standard OpenAI API integration. The implementation maintains consistency with the existing architecture while adding flexibility in model selection and configuration.

## Key Changes
- Added new OpenAI provider with support for multiple models (gpt-4o, o1, o3-mini, gpt-4-turbo)
- Implemented environment variable configuration for OpenAI API key and model selection
- Created comprehensive test suite covering:
  - Basic connectivity and configuration
  - Tool binding and usage
  - Error handling and rate limiting
  - Multi-turn dialogues
  - Model variant compatibility
- Updated documentation with OpenAI setup instructions and usage examples

## Technical Details
- Uses LangChain's built-in tool handling for consistent agent behavior
- Implements proper error handling for API key and model validation
- Maintains backward compatibility with existing Azure OpenAI integration
- Adds support for environment variable overrides with sensible defaults

## Testing
- Added unit tests with mocked API calls
- Added integration tests with real API calls (skipped if no API key)
- All tests passing with proper error handling and edge cases covered

## Documentation
- Added OpenAI-specific configuration section to README
- Documented supported models and their use cases
- Included example usage with direct OpenAI integration
- Updated environment variable documentation

## Dependencies
- No new dependencies required beyond existing LangChain integration

## Notes
- Default model is set to `gpt-4o` for optimal agent interactions
- Azure OpenAI functionality remains unchanged
- Added proper warning handling for Pydantic deprecation notices

### CODE REVIEW

**Comment by eleonorecharles** [comment]
Hi @EdwardTang,

Thank you for this PR! OpenAI integration is a great addition.

We have just merged a PR (https://github.com/datalayer/jupyter-ai-agents/pull/15) that adds `--model-provider` and `--model-name` fields.

Once you adjust your PR to match these changes, you can undraft it, and we will merge.

**Comment by EdwardTang** [comment]
> Hi @EdwardTang,
> 
> Thank you for this PR! OpenAI integration is a great addition.
> 
> We have just merged a PR (#15) that adds `--model-provider` and `--model-name` fields.
> 
> Once you adjust your PR to match these changes, you can undraft it, and we will merge.

Awesome! let me do that real quick

File: jupyter_ai_agents/base.py [hunk 1]
```diff
@@ -89,12 +89,7 @@ class JupyterAIAgentBaseApp(JupyterApp):
     model_provider = Unicode(
         "github-copilot",
         config=True,
-        help="Model provider can be 'azure-openai' or 'github-copilot'."
-    )
-    openai_api_version = Unicode(
```
Comment (eleonorecharles) [inline jupyter_ai_agents/base.py:94]: same as above

File: jupyter_ai_agents/base.py [hunk 2]
```diff
@@ -41,9 +41,9 @@
         "agent": "JupyterAIAgentBaseApp.agent_name",
         "input": "JupyterAIAgentBaseApp.input",
         "model-provider": "JupyterAIAgentBaseApp.model_provider",
-        "openai-api-version": "JupyterAIAgentBaseApp.openai_api_version",
```
Comment (eleonorecharles) [inline jupyter_ai_agents/base.py:44]: @EdwardTang could you not delete "openai_api_version"? It is required for azure openai

File: README.md [hunk 3]
```diff
@@ -85,6 +85,34 @@ AZURE_OPENAI_API_KEY="..."
 EOF
 ```
 
+### Using Direct OpenAI Integration
```
Comment (eleonorecharles) [inline README.md:88]: Could you move this to a new ```openai``` folder under ```docs/docs/models``` ?

**Comment by norahsakal** [comment]
Hi @EdwardTang, thanks for your awesome work on this PR! I've cloned and tested the branch locally, and it works great. Just a friendly nudge - excited to see this merged! 🚀

File: README.md (hunk 3 above)
Comment (EdwardTang) [reply README.md:88]: Sure!

File: jupyter_ai_agents/base.py (hunk 2 above)
Comment (EdwardTang) [reply jupyter_ai_agents/base.py:44]: Will do.

**Comment by EdwardTang** [comment]
@eleonorecharles thank you for your comments! sorry for the delay due to vacation. please take a look and let me know if they looks good now.

**Comment by EdwardTang** [comment]
> @EdwardTang We have just moved the base.py file under the cli folder... which may give you trouble for rebase... hopefully not too much.
> 
> There is also now a need to ensure you don't have pycrdt package, but rather datalayer_pycrdt (bump also the jupyter-nbmodel-client). That is temporary and we are working on a more robust solution.

@echarles  could you please review this PR Thanks a lot!

**Review by echarles — APPROVED** [review]
LGTM Thx @EdwardTang
